L1Trigger/DemonstratorTools/plugins/GTTInputFileWriter.cc

```diff
@@ -0,0 +1,177 @@
+
+    {9, {kTrackTMUX, 0, kGapLength}},   {10, {kTrackTMUX, 0, kGapLength}},  {11, {kTrackTMUX, 0, kGapLength}},
+    {12, {kTrackTMUX, 0, kGapLength}},  {13, {kTrackTMUX, 0, kGapLength}},  {14, {kTrackTMUX, 0, kGapLength}},
+    {15, {kTrackTMUX, 0, kGapLength}},  {16, {kTrackTMUX, 0, kGapLength}},  {17, {kTrackTMUX, 0, kGapLength}},
+
+    {18, {kTrackTMUX, 6, kGapLength}},  {19, {kTrackTMUX, 6, kGapLength}},  {20, {kTrackTMUX, 6, kGapLength}},
+    {21, {kTrackTMUX, 6, kGapLength}},  {22, {kTrackTMUX, 6, kGapLength}},  {23, {kTrackTMUX, 6, kGapLength}},
+    {24, {kTrackTMUX, 6, kGapLength}},  {25, {kTrackTMUX, 6, kGapLength}},  {26, {kTrackTMUX, 6, kGapLength}},
+
+    {27, {kTrackTMUX, 6, kGapLength}},  {28, {kTrackTMUX, 6, kGapLength}},  {29, {kTrackTMUX, 6, kGapLength}},
+    {30, {kTrackTMUX, 6, kGapLength}},  {31, {kTrackTMUX, 6, kGapLength}},  {32, {kTrackTMUX, 6, kGapLength}},
+    {33, {kTrackTMUX, 6, kGapLength}},  {34, {kTrackTMUX, 6, kGapLength}},  {35, {kTrackTMUX, 6, kGapLength}},
+
+    {36, {kTrackTMUX, 12, kGapLength}}, {37, {kTrackTMUX, 12, kGapLength}}, {38, {kTrackTMUX, 12, kGapLength}},
+    {39, {kTrackTMUX, 12, kGapLength}}, {40, {kTrackTMUX, 12, kGapLength}}, {41, {kTrackTMUX, 12, kGapLength}},
+    {42, {kTrackTMUX, 12, kGapLength}}, {43, {kTrackTMUX, 12, kGapLength}}, {44, {kTrackTMUX, 12, kGapLength}},
+
+    {45, {kTrackTMUX, 12, kGapLength}}, {46, {kTrackTMUX, 12, kGapLength}}, {47, {kTrackTMUX, 12, kGapLength}},
+    {48, {kTrackTMUX, 12, kGapLength}}, {49, {kTrackTMUX, 12, kGapLength}}, {50, {kTrackTMUX, 12, kGapLength}},
+    {51, {kTrackTMUX, 12, kGapLength}}, {52, {kTrackTMUX, 12, kGapLength}}, {53, {kTrackTMUX, 12, kGapLength}}};
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+
+GTTInputFileWriter::GTTInputFileWriter(const edm::ParameterSet& iConfig)
+    : tracksToken_(consumes<edm::View<Track_t>>(iConfig.getUntrackedParameter<edm::InputTag>("tracks"))),
+      eventCount_(0),
+      fileWriter_(l1t::demo::parseFileFormat(iConfig.getUntrackedParameter<std::string>("format")),
+                  iConfig.getUntrackedParameter<std::string>("outputFilename"),
+                  9,
+                  6,
+                  1024,
+                  kChannelSpecs) {
+  //now do what ever initialization is needed
+}
+
+GTTInputFileWriter::~GTTInputFileWriter() {
+  // do anything here that needs to be done at desctruction time
+  // (e.g. close files, deallocate resources etc.)
+  //
+  // please remove this method altogether if it would be left empty
+}
+
+void GTTInputFileWriter::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+
+  // Select links for correct time slice (18 input links per time slice)
+  // TODO: Find nicer solution - e.g. having BoardDataWriter automatically shuffle
+  //       data between channels based on event-independent 'logical' ID
+  //       (at least nicer solution than storing event count as member var)
```

Comment:
some thoughts that I had sent to Tom by email: for a case where a board at TMUX N(6) receives M input links each from K(3) time-slices of a system at TMUX K&times;N(18), one could have an adapter class that:
 * internally holds a counter from 0 to K-1
 * receives from the user at each event the data on M channels
 * writes those out to K&times;M channels in the BoardData (and can create the config for the K&times;M channels from the config for the M)
IMHO it would be good to also support both the mode where the links of each timeslice are kept together (K groups of M links) and the mode where they're interleaved (i.e. M groups of K links), as both could be viable options in the the real system depending on how the firmware and patch panels will be implemented

Reply:
Thanks for your comments. I agree that the API needs to support cases where the links from each time slice are in contiguous blocks (i.e. slice 0 arrives on links X to links X+17, slice 1 on links X+18 to X+35, etc) as well as cases where the links are interleaved in a repeated pattern. Beyond that, I also want to retain the ability in the API to also support breaks in such patterns (e.g. in case links from one upstream system need to arrive in the same SLR on both sides of the chip, or in case links for certain regions need to be mapped onto different SLRs).

I've now pushed a couple of new commits that update the `BoardDataReader` & `BoardDataWriter` implementations to use these 'event-index-independent' link IDs. To support the full range of possible use cases, the new `BoardDataReader` & `BoardDataWriter` API allows for any mapping of 'event-index-independent' link IDs to board framework channel indices. With this API, for any specific subsystem whose inputs & output channels have the nice repeated patterns that you describe, it should also be possible to generate those maps (or parts of them) from a function with a small number of parameters.